src/record_syscall.cc

```diff
@@ -3783,8 +3783,9 @@ static void process_mremap(RecordTask* t, remote_ptr<void> old_addr,
     return;
   }
 
-  // Now record the new part of the mapping.
-  km = m.map.subrange(new_addr + old_size, new_addr + new_size);
+  // Now record the new part of the mapping, possibly with some overlap.
+  km = m.map.subrange(new_addr + floor_page_size(old_length),
```

Comment:
How can `st.st_size` be somewhere below that? `new_length` must be >= `page_size * (N + 1)`. It sounds like this must be a case where we're mapping one or more pages entirely beyond the current EOF? That does look like it would be a problem here ... but then I think the correct fix would simply be to max `end` with zero.